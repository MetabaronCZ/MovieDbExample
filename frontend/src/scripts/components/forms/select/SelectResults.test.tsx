import { act } from '@testing-library/react';
import { describe, expect, it, jest } from '@jest/globals';
import {
  fireEvent,
  getAllByRole,
  getByRole,
  getByTestId,
  queryByRole,
} from '@testing-library/dom';

import { SelectResults } from 'components/forms/select/SelectResults';
import { TestComponentWrapper, waitForComponent } from 'test-utils/component';

// mock Loader component
jest.mock('../../common/Loader.tsx', () => ({
  Loader: () => <div data-testid="loader-mock" />,
}));

describe('components/forms/select/SelectResults', () => {
  it('should render properly', async () => {
    const { container } = await waitForComponent(
      <SelectResults
        options={[
          { title: 'Item 1', value: 1 },
          { title: 'Item 2', value: 2 },
        ]}
        onSelect={() => null}
      />,
      { wrapper: TestComponentWrapper },
    );
    expect(container).toMatchSnapshot();

    const list = getByRole(container, 'list');
    expect(list).toBeInTheDocument();

    const items = getAllByRole(container, 'listitem');
    expect(items.length).toEqual(2);
    expect(items[0]).toHaveTextContent('Item 1');
    expect(items[1]).toHaveTextContent('Item 2');
  });

  it('should render result item props', async () => {
    const { container } = await waitForComponent(
      <SelectResults
        options={[
          {
            title: 'Item',
            value: 101,
            extra: <sub>EXTRA</sub>,
            description: 'DESCRIPTION',
            disabled: true,
          },
        ]}
        onSelect={() => null}
      />,
      { wrapper: TestComponentWrapper },
    );
    expect(container).toMatchSnapshot();

    const item = getByRole(container, 'listitem');
    expect(item).toBeInTheDocument();

    const button = item.querySelector('button');
    expect(button).toBeInTheDocument();

    expect(button).toHaveAttribute('title', 'DESCRIPTION');
    expect(button).toHaveAttribute('disabled');
    expect(button).toHaveTextContent('ItemEXTRA');
  });

  it('should render empty options', async () => {
    const { container } = await waitForComponent(
      <SelectResults options={[]} onSelect={() => null} />,
      { wrapper: TestComponentWrapper },
    );
    expect(container).toMatchSnapshot();
    expect(container).toHaveTextContent('Nebyla nalezena žádná data.');

    const list = queryByRole(container, 'list');
    expect(list).not.toBeInTheDocument();
  });

  it('should render left-aligned variant', async () => {
    const { container } = await waitForComponent(
      <SelectResults
        align="left"
        options={[
          { title: 'Item 1', value: 1 },
          { title: 'Item 2', value: 2 },
        ]}
        onSelect={() => null}
      />,
      { wrapper: TestComponentWrapper },
    );
    expect(container).toMatchSnapshot();
  });

  it('should render right-aligned variant', async () => {
    const { container } = await waitForComponent(
      <SelectResults
        align="right"
        options={[
          { title: 'Item 1', value: 1 },
          { title: 'Item 2', value: 2 },
        ]}
        onSelect={() => null}
      />,
      { wrapper: TestComponentWrapper },
    );
    expect(container).toMatchSnapshot();
  });

  it('should render loading state', async () => {
    const { container } = await waitForComponent(
      <SelectResults
        options={[
          { title: 'Item 1', value: 1 },
          { title: 'Item 2', value: 2 },
        ]}
        loading
        onSelect={() => null}
      />,
      { wrapper: TestComponentWrapper },
    );
    expect(container).toMatchSnapshot();
    expect(container).toHaveTextContent('');

    const list = queryByRole(container, 'list');
    expect(list).not.toBeInTheDocument();

    // Loader present
    const loader = getByTestId(container, 'loader-mock');
    expect(loader).toBeInTheDocument();
  });

  it('should select items', async () => {
    const onSelect = jest.fn();

    const { container } = await waitForComponent(
      <SelectResults
        options={[
          { title: 'Item 1', value: 1 },
          { title: 'Item 2', value: 2 },
        ]}
        onSelect={onSelect}
      />,
      { wrapper: TestComponentWrapper },
    );
    expect(onSelect).not.toHaveBeenCalled();

    const list = getByRole(container, 'list');
    const items = getAllByRole(list, 'button');

    // click Item 2
    act(() => {
      fireEvent.click(items[1]);
    });
    expect(onSelect).toHaveBeenCalledTimes(1);
    expect(onSelect).toHaveBeenCalledWith(2); // Item 2 value
  });

  it('should not select disabled item', async () => {
    const onSelect = jest.fn();

    const { container } = await waitForComponent(
      <SelectResults
        options={[
          { title: 'Item 1', value: 1 },
          { title: 'Item 2', value: 2, disabled: true },
        ]}
        onSelect={onSelect}
      />,
      { wrapper: TestComponentWrapper },
    );
    expect(onSelect).not.toHaveBeenCalled();

    const list = getByRole(container, 'list');
    const items = getAllByRole(list, 'button');

    // click Item 2
    act(() => {
      fireEvent.click(items[1]);
    });
    expect(onSelect).not.toHaveBeenCalled();
  });
});
