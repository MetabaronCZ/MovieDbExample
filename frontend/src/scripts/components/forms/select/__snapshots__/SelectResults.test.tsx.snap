// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`components/forms/select/SelectResults should render empty options 1`] = `
.c0 {
  position: absolute;
  top: calc(100% + 4px);
  left: 0;
  width: 240px;
  overflow: hidden;
  z-index: 1000;
  background: #fff;
  border: 1px solid #002d56;
  box-shadow: 1px 1px 4px #dae0e3;
}

.c1 {
  font-family: Tahoma,Verdana,Arial,sans-serif;
  font-size: 14px;
  line-height: 24px;
  color: #002d56;
  padding: 8px;
  font-style: italic;
  text-align: center;
}

<div>
  <div
    class="c0"
  >
    <div
      class="c1"
    >
      Nebyla nalezena žádná data.
    </div>
  </div>
</div>
`;

exports[`components/forms/select/SelectResults should render left-aligned variant 1`] = `
.c3 {
  display: block;
  border: none;
  background: transparent;
  outline: none;
  cursor: pointer;
}

.c3:disabled {
  cursor: not-allowed;
}

.c0 {
  position: absolute;
  top: calc(100% + 4px);
  right: 0;
  width: 240px;
  overflow: hidden;
  z-index: 1000;
  background: #fff;
  border: 1px solid #002d56;
  box-shadow: 1px 1px 4px #dae0e3;
}

.c1 {
  list-style-type: none;
  overflow-y: auto;
  max-height: 360px;
}

.c2 {
  border-bottom: 1px solid #dae0e3;
}

.c2:last-child {
  border-bottom: none;
}

.c4 {
  font-family: Tahoma,Verdana,Arial,sans-serif;
  font-size: 14px;
  line-height: 24px;
  color: #002d56;
  width: 100%;
  padding: 4px 8px;
  background: #fff;
  white-space: nowrap;
  text-align: left;
}

.c4:hover {
  background-color: #f2f7fa;
}

.c4:focus {
  outline: 1px dashed #002d56;
  outline-offset: -1px;
}

.c4:disabled {
  color: #98c3d9;
}

<div>
  <div
    class="c0"
  >
    <ul
      class="c1"
    >
      <li
        class="c2"
      >
        <button
          class="c3 c4"
          type="button"
        >
          Item 1
        </button>
      </li>
      <li
        class="c2"
      >
        <button
          class="c3 c4"
          type="button"
        >
          Item 2
        </button>
      </li>
    </ul>
  </div>
</div>
`;

exports[`components/forms/select/SelectResults should render loading state 1`] = `
.c0 {
  position: absolute;
  top: calc(100% + 4px);
  left: 0;
  width: 240px;
  overflow: hidden;
  z-index: 1000;
  background: #fff;
  border: 1px solid #002d56;
  box-shadow: 1px 1px 4px #dae0e3;
}

<div>
  <div
    class="c0"
  >
    <div
      data-testid="loader-mock"
    />
  </div>
</div>
`;

exports[`components/forms/select/SelectResults should render properly 1`] = `
.c3 {
  display: block;
  border: none;
  background: transparent;
  outline: none;
  cursor: pointer;
}

.c3:disabled {
  cursor: not-allowed;
}

.c0 {
  position: absolute;
  top: calc(100% + 4px);
  left: 0;
  width: 240px;
  overflow: hidden;
  z-index: 1000;
  background: #fff;
  border: 1px solid #002d56;
  box-shadow: 1px 1px 4px #dae0e3;
}

.c1 {
  list-style-type: none;
  overflow-y: auto;
  max-height: 360px;
}

.c2 {
  border-bottom: 1px solid #dae0e3;
}

.c2:last-child {
  border-bottom: none;
}

.c4 {
  font-family: Tahoma,Verdana,Arial,sans-serif;
  font-size: 14px;
  line-height: 24px;
  color: #002d56;
  width: 100%;
  padding: 4px 8px;
  background: #fff;
  white-space: nowrap;
  text-align: left;
}

.c4:hover {
  background-color: #f2f7fa;
}

.c4:focus {
  outline: 1px dashed #002d56;
  outline-offset: -1px;
}

.c4:disabled {
  color: #98c3d9;
}

<div>
  <div
    class="c0"
  >
    <ul
      class="c1"
    >
      <li
        class="c2"
      >
        <button
          class="c3 c4"
          type="button"
        >
          Item 1
        </button>
      </li>
      <li
        class="c2"
      >
        <button
          class="c3 c4"
          type="button"
        >
          Item 2
        </button>
      </li>
    </ul>
  </div>
</div>
`;

exports[`components/forms/select/SelectResults should render result item props 1`] = `
.c3 {
  display: block;
  border: none;
  background: transparent;
  outline: none;
  cursor: pointer;
}

.c3:disabled {
  cursor: not-allowed;
}

.c0 {
  position: absolute;
  top: calc(100% + 4px);
  left: 0;
  width: 240px;
  overflow: hidden;
  z-index: 1000;
  background: #fff;
  border: 1px solid #002d56;
  box-shadow: 1px 1px 4px #dae0e3;
}

.c1 {
  list-style-type: none;
  overflow-y: auto;
  max-height: 360px;
}

.c2 {
  border-bottom: 1px solid #dae0e3;
}

.c2:last-child {
  border-bottom: none;
}

.c4 {
  font-family: Tahoma,Verdana,Arial,sans-serif;
  font-size: 14px;
  line-height: 24px;
  color: #002d56;
  width: 100%;
  padding: 4px 8px;
  background: #fff;
  white-space: nowrap;
  text-align: left;
}

.c4:hover {
  background-color: #f2f7fa;
}

.c4:focus {
  outline: 1px dashed #002d56;
  outline-offset: -1px;
}

.c4:disabled {
  color: #98c3d9;
}

<div>
  <div
    class="c0"
  >
    <ul
      class="c1"
    >
      <li
        class="c2"
      >
        <button
          class="c3 c4"
          disabled=""
          title="DESCRIPTION"
          type="button"
        >
          Item
          <sub>
            EXTRA
          </sub>
        </button>
      </li>
    </ul>
  </div>
</div>
`;

exports[`components/forms/select/SelectResults should render right-aligned variant 1`] = `
.c3 {
  display: block;
  border: none;
  background: transparent;
  outline: none;
  cursor: pointer;
}

.c3:disabled {
  cursor: not-allowed;
}

.c0 {
  position: absolute;
  top: calc(100% + 4px);
  left: 0;
  width: 240px;
  overflow: hidden;
  z-index: 1000;
  background: #fff;
  border: 1px solid #002d56;
  box-shadow: 1px 1px 4px #dae0e3;
}

.c1 {
  list-style-type: none;
  overflow-y: auto;
  max-height: 360px;
}

.c2 {
  border-bottom: 1px solid #dae0e3;
}

.c2:last-child {
  border-bottom: none;
}

.c4 {
  font-family: Tahoma,Verdana,Arial,sans-serif;
  font-size: 14px;
  line-height: 24px;
  color: #002d56;
  width: 100%;
  padding: 4px 8px;
  background: #fff;
  white-space: nowrap;
  text-align: left;
}

.c4:hover {
  background-color: #f2f7fa;
}

.c4:focus {
  outline: 1px dashed #002d56;
  outline-offset: -1px;
}

.c4:disabled {
  color: #98c3d9;
}

<div>
  <div
    class="c0"
  >
    <ul
      class="c1"
    >
      <li
        class="c2"
      >
        <button
          class="c3 c4"
          type="button"
        >
          Item 1
        </button>
      </li>
      <li
        class="c2"
      >
        <button
          class="c3 c4"
          type="button"
        >
          Item 2
        </button>
      </li>
    </ul>
  </div>
</div>
`;
